### PR TITLE
Improvements to managed_ptr

### PR DESCRIPTION
**Pull request by adayton1**
Adds chai::unpack so that the user can specify when they want a ManagedArray or a managed_ptr to be unpacked when calling chai::make_managed. If chai::unpack is used, the host pointer will be extracted and passed to the host constructor of the object and the device pointer will be extracted and passed to the device constructor of the object. The current behavior of automatically detecting when to do this falls down in the case where some ManagedArrays/managed_ptrs should be unpacked and others should not be, particularly when implicit conversions are disabled.

### CODE REVIEW

**Comment by robinson96** [comment]
Can we update sphinx documentation to explain this stuff?

File: src/chai/managed_ptr.hpp [hunk 1]
```diff
@@ -601,7 +609,80 @@ namespace chai {
          }
    };
 
+   ///
+   /// @author Alan Dayton
+   ///
+   /// A wrapper used for unpacking the internal pointers in the correct space
```
[the remark this answers is not in the document]
Comment (adayton1) [reply src/chai/managed_ptr.hpp]: Added a bit more. I'll add more detail in the sphinx documentation.

**Comment by adayton1** [comment]
I've added sphinx documentation to explain about managed_ptr.

**Review by davidbeckingsale — APPROVED** [review]
Looks good to me. Thanks for the documentation.

**Review by robinson96 — APPROVED** [review]
LGTM

**Comment by davidbeckingsale** [comment]
LGTM.

**Pull request merged**